### PR TITLE
Fixes MatrixFilterContents rendering/coverage

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes: https://github.com/flutter/flutter/issues/147807
relands https://github.com/flutter/engine/pull/43943 (with fixes that hopefully avoid it being reverted again)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/entity/contents/filters/matrix_filter_contents.cc [hunk 1]
```diff
@@ -28,6 +28,17 @@ void MatrixFilterContents::SetSamplerDescriptor(SamplerDescriptor desc) {
   sampler_descriptor_ = std::move(desc);
 }
 
+namespace {
+Matrix CalculateSubpassTransform(const Matrix& entity_transform,
```
Comment (gaaclarke) [inline impeller/entity/contents/filters/matrix_filter_contents.cc:32]: @bdero this is what I was talking about the other day.  We should be sharing the math between rendering and coverage.

**Review by bdero — APPROVED** [review]
LGTM! Thanks for looking through this.

**Comment by gaaclarke** [comment]
The PR has some unexpected golden changes I'll have to audit:

## example
Screenshot 2024-05-16 at 3 54 56 PM

**Comment by gaaclarke** [comment]
Here is the above test animated, it definitely doesn't look right with this patch.  It is hitting the subpass code.  It looks like that AiksTest was intended to duplicate case we are trying to fix.  I'm not sure how it differs yet (beyond one being written at aiks and the other being written at dl).

scale

**Comment by gaaclarke** [comment]
The difference between `RenderCoverageMatchesGetCoverageSubpassScale` (which works) and `MatrixImageFilterMagnify` (which doesn't) is `MatrixImageFilterMagnify` has a translation in the `effect_transform` whereas the other test just has a scalar.

- ❌ `MatrixImageFilterMagnify`
  - snapshot->transform = Translate(1200, -200)
  - effect_transform = Translate(1200, -400) * Scale(2,2,1)
   - matrix_ = Scale(2,2,1)
- ✅ `RenderCoverageMatchesGetCoverageSubpassScale`
  - snapshot->transform = Translate(100, 200)
  - effect_transform = Scale(2,2,1)
  - matrix_ = Scale(3,3,1)
- ✅ `Bug147807`
  - snapshot->transform = Translate(-402.5, -20)
  - effect_transform = Scale(2, 2, 1)
  - matrix_ = Translate(-280, -920) * Scale(3,3,1)

Having a translation in the `effect_transform` doesn't match what I'm seeing when using the reproduction app or the `Bug147807` bug which duplicates it.

**Comment by gaaclarke** [comment]
This passes all tests, but @bdero and I are auditing it to make sure it's correct.  So don't land this yet.

**Comment by gaaclarke** [comment]
I think the check for translation hack works because it is by proxy detecting the Canvas::Save flow as opposed to the ImageFilter flow.  In the Canvas::Save path the effect_transform is truncated to its basis before it is passed down ( https://github.com/flutter/engine/blob/4e33c102e56e9b3e9898c93c06373066c7eefc92/impeller/entity/entity_pass.cc#L590 )

**Comment by gaaclarke** [comment]
@bdero I made that change we discussed, can you please take a look.

We decided offline that we are clearly seeing 3 different cases that need to be handled at the MatrixImageFilterContents level.  So we expanded the RenderMode to reflect those 3 cases.  I suspect maybe we are misusing the effect transform where we get it meaning 3 different things in 3 different contexts.  Hopefully we can clear that up in the future.

**Review by bdero — APPROVED** [review]
LGTM, thanks!

**Comment by jonahwilliams** [comment]
This looks like its working incorrectly with some magnifier goldens in the framework:

https://flutter-gold.skia.org/search?crs=github&issue=148593&left_filter=name%3Dwidgets.widgets.magnifier.styled&master=true&patchsets=3&positive=true

Top is the Impeller golden diff and bottom is current Skia (which is unchanged)

image

**Comment by jonahwilliams** [comment]
reason for revert: unexpected framework golden change

**Comment by gaaclarke** [comment]
Here are the matrices for the magnifier.styled test:
- snapshot: translate(-600,-600)
- effect: scale(3,3,1)
- matrix: translate(-150, -150) * scale(2, 2, 1)

It's going through the `kBackdropSubpass` flow [here](https://github.com/flutter/engine/pull/52880/files#diff-d2df72e67028e1f1ed7f094a7a5427fa0dfb0ff4e9aab6eb6dbc1980ae1ed6d8L590).  Backing out the backdrop flow fixes all tests except `Bug147807` (the current state of the code.

All engine and frameworks tests pass with the following patch:
```diff
diff --git a/impeller/entity/entity_pass.cc b/impeller/entity/entity_pass.cc
index ee28a0dd4f..1cc997d5cc 100644
--- a/impeller/entity/entity_pass.cc
+++ b/impeller/entity/entity_pass.cc
@@ -585,9 +586,10 @@ EntityPass::EntityResult EntityPass::GetEntityForElement(
       auto texture = pass_context.GetTexture();
       // Render the backdrop texture before any of the pass elements.
       const auto& proc = subpass->backdrop_filter_proc_;
       subpass_backdrop_filter_contents = proc(
           FilterInput::Make(std::move(texture)), subpass->transform_.Basis(),
-          Entity::RenderingMode::kBackdropSubpass);
+          Entity::RenderingMode::kImageFilterSubpass);
```

So the concept is right, there is a third state, the `EntityPass::GetElementsCoverage` flow for backdrop filters isn't passing in the basis of the transform, whereas the `EntityPass::GetEntityForElement` is.  That appears to be the detection we need to differentiate the different cases.  I plan on relanding this with the above patch and more apropos names for the different math than `BackdropSubpass vs ImageFilterSubpass`.

**Comment by gaaclarke** [comment]
As an experiment I tried to remove the `Basis()` call in EntityPass::GetEntityForElement ([here](https://github.com/flutter/engine/blob/8eba63f29b8b26931496807cb4fc2fed8e4e3258/impeller/entity/entity_pass.cc#L590)) with Brandon's original patch.  It get's the same results as this PR, all the engine tests pass but the framework one fails.  It seems like we still need that (barring some sort of deeper refactoring).

**Comment by gaaclarke** [comment]
This patch seems to make the engine and framework tests happy.  I'm not familiar with ContentBoundsPromise but it roughly sounds like it could effect the contents we get from the snapshot and determines how we should be performing the math:
```diff
diff --git a/impeller/entity/entity_pass.cc b/impeller/entity/entity_pass.cc
index 984403629c..13e99b3ced 100644
--- a/impeller/entity/entity_pass.cc
+++ b/impeller/entity/entity_pass.cc
@@ -585,9 +585,17 @@ EntityPass::EntityResult EntityPass::GetEntityForElement(
       auto texture = pass_context.GetTexture();
       // Render the backdrop texture before any of the pass elements.
       const auto& proc = subpass->backdrop_filter_proc_;
       subpass_backdrop_filter_contents = proc(
           FilterInput::Make(std::move(texture)), subpass->transform_.Basis(),
-          Entity::RenderingMode::kImageFilterSubpass);
+          subpass->bounds_promise_ == ContentBoundsPromise::kContainsContents ?
+          Entity::RenderingMode::kImageFilterSubpass :
+          Entity::RenderingMode::kBackdropSubpass);
 
       // If the very first thing we render in this EntityPass is a subpass that
       // happens to have a backdrop filter, than that backdrop filter will end
```

If this is the correct latch we can rename the render modes accordingly.

**Pull request merged**